### ISSUE TITLE
Review and finalize options

### ISSUE THREAD

**Issue by dseidman** [BrewTroller/BrewTroller-Official#20]
There are many, many options in config.in.h, some of which seem very random (e.g. MASH_PREHEAT_NO_VALVES), and some of which aren't implemented or don't work (batch sparging, auto chill). We should finalize the list of the ones we think should continue to be supported, and of those, which should be configured at runtime vs. compile time.  For example, why is PWM a compile-time option but PID a runtime option? There may be a good reason but it isn't obvious.

**Comment by ericyanush** [comment]
I agree; there are a massive amount of options, most have questionable use (the 80-20 rule applies). The list of options to be officially supported needs to be clearly defined and documented.

**Comment by dseidman** [comment]
Okay, here is my analysis of the options. The data from the survey may change this, but probably not by much. Also, I think ultimately we should move towards a system definition model that has the user define the way that vessels in their system are connected together and what type of heat they use. This would supercede a number of options that are currently set manually. There are still about 25 options, but I don't think any of them can easily be cut without breaking a substantial portion of systems. A bunch of these options are also used to fix problems that can occur, so we could also remove them from the default UI but add a "troubleshooting" section that exposes them and provides examples of when to use them.

Keep as compile-time options:
#cmakedefine USEMETRIC
#define GRAIN_VOL_LOSS 1.7884
#define PIDLIMIT_HLT 100 We might be able to cut these, I'm not sure what the scenario is to use a limit except for the boil (boilover, caramelization) and pump (compacting the mash, splashing).
#define PIDLIMIT_MASH 100
#define PIDLIMIT_KETTLE 100
#define PIDLIMIT_STEAM 100 
#define PIDLIMIT_PUMP 100
#define FEEDFORWARD ${FF_OPT} Should better document how to set this up if we keep it
#define FLOWRATE_READ_INTERVAL
#define PID_FLOW_MIN 30
#define SMART_HERMS_HLT We might want to just consolidate all the SMART_HERMS code into a single HERMS option
#define MASH_HEAT_LOSS 0
#define SMART_HERMS_PREHEAT
#define STRIKE_TEMP_OFFSET 0
#cmakedefine MASH_AVG_AUX1 
#cmakedefine MASH_AVG_AUX2 
#cmakedefine MASH_AVG_AUX3 
All of the AUTO_ config options
#define RIMS_MLT_SETPOINT_DELAY However, we should fix the code so that it works right even if the program doesn’t have a dough-in step longer than this delay. It should not be possible to create a config that breaks the software.
#define RIMS_DELAY 60000
#cmakedefine VOLUME_MANUAL
#define VOLUME_READ_INTERVAL 200
#define VOLUME_READ_COUNT 5
Modbus if the survey shows people are using it
#define HLT_MIN_HEAT_VOL 4000 No idea why these are marked experimental, the code is very simple
#define MASH_MIN_HEAT_VOL 4000
#define KETTLE_MIN_HEAT_VOL 4000
#define STEAM_OFFSET 5 If anyone is actually using steam
#define RIMS_DURING_SPARGE
All the RGBIO options, it seems like these are widely used

Move to UI:
#define HOPADD_DELAY 0 If you’ve ever tried tuning a hop carousel’s timing using this option, you’ll know why I want it in the UI… 

Remove, requires effort to support:
#define MASH_PREHEAT_SENSOR TS_AUX1
#define MASH_PREHEAT_STRIKE (note that these are really easy to support if there’s any demand)
#define MASH_PREHEAT_STEP1
#define PID_CONTROL_MANUAL
#define HLT_FLY_SPARGE_STOP
#define HLT_FLY_SPARGE_STOP_VOLUME 250
Everything related to Serial Com, unless the survey shows people using it still. Even then, we should cut the old COM schemas and only support the latest.
Everything related to BTPD, seems like few/nobody are using these?
#define HLT_MIN_REFILL
#define HLT_MIN_REFILL_VOL 4000
#define BATCH_SPARGE This was never implemented and would be a pretty big pain to support, and I have no idea why you would want to do it when fly sparge is automated.
#define BATCH_VOLUME_OFFSET 0
#define BATCH_SPARGE_RECIRC 60
#define HLT_MAX_TEMP 180

Implement better, make default and remove option – there is no reason to turn off the HLT setpoint during sparge when we have multiple protections against dry fire:
#define HLT_HEAT_SPARGE

Remove because HLT_MIN_HEAT_VOLUME does the same thing:
#define HLT_MIN_SPARGE 2000

Remove, easy to support, but clutter the UI:
#define KETTLELID_THRESH 80 
#define GRAIN2VOL 1.25
#define VOL_SHRINKAGE 1
#define BOIL_OFF_GALLONS
#define PREBOIL_ALARM 205 And I even use this… 
#define BUZZER_CYCLE_TIME 1200 We should switch the default to a modulated buzzer, steady buzzer is really annoying 
#define BUZZER_ON_TIME 500  
#define LOGO_BREWTROLLER

Newly defined by me:
Enable smoothing of flowrates:
#define FLOWRATE_SAMPLES 

In place of previous opaque PWM options:
#cmakedefine USEPWM_HLT
#cmakedefine USEPWM_MASH
#cmakedefine USEPWM_KETTLE
#cmakedefine USEPWM_STEAM
#cmakedefine USEPWM_PUMP1
#cmakedefine USEPWM_PUMP2

To support new PID pump functionality:
#define PID_PUMP1 
#define PID_PUMP2

Discussed separately:
#cmakedefine CUSTOM_NAME_1 "Your Brewery"

Refactor – the user should explicitly specify the number of vessels and the role that each one plays, such as NUM_VESSELS=2, VS_HLT = VESSEL_1, VS_MASH = VESSEL_1, VS_KETTLE = VESSEL_2
#cmakedefine HLT_AS_KETTLE
#cmakedefine KETTLE_AS_MASH
#cmakedefine SINGLE_VESSEL_SUPPORT

Move to another file:
constexpr int HEAT_OUTPUTS_COUNT = 4;
constexpr uint8_t NUM_VESSELS = HEAT_OUTPUTS_COUNT;

I don’t understand the purpose of these:
static const uint8_t VS = 0;
static const uint8_t TS = 1;

**Comment by ericyanush** [comment]
Based on the current responses to the user survey (If I can find a way to automatically generate charts with anonymized info, I will post a link to them) a **_large_** portion of the options aren't broadly used, and can likely be removed.

### General Summary of responses

##### Hardware
- Majority of users have DX1/2 or 4.x boards
- 75% of users do not have modbus boards
- 50% of Users have RGBIO boards
- 75% have I2CLCD
- 66% have BTNIC
- 33% have PID displays
- 33% have E-Stop

##### Basic Options
- 75% use Imperial Units
- 70% don't change the brewing calculation factors (GRAIN2VOL, SHRINKAGE, etc)
- 60% don't enable any vessel options (HLT_AS_KETTLE, SINGLE_VESSEL)
- 85% of users don't use the PID output power limit configs
- 75% of users don't use PID forward feed options
- 75% of users don't enable timer PWM
- 85% of users don't use PID_FLOW_CONTROL or PID_CONTROL_MANUAL
- 62% of users don't enable Sparge in Pump Control
- 100% of users don't change Kettle Lid Threshold
- 93% of users don't change the Hop add delay
- 75% of users don't use Smart Herms
- 50% of users change the Strike Temperature offset
- 60% of users don't enable Mash AUX temperature sensor averaging.
- 70% of users don't enable HLT fly sparge stop
- 70% of users don't change the Boil off units (boil off gallons)
- 70% of users don't change the Pre-boil alarm value
- 70% of users don't use Buzzer modulation
- 75% of users use at least one Brew step automation
- 60% of users don't enable MLT set point delay.
- 93% of users don't enable volume manual.
- 60% of users don't change the volume sensor averaging parameters
- 83% of users don't enable flow rate calculations

##### Experimental Options
- 50% of users use Element saver options
- 50% of users enable HLT minimum value refill
- 50% of users enable HLT heat during sparge
- **_100% of users don't use the direct fired RIMS support_**
- 75% of users don't enable Batch Sparge support
- _70% of users don't enable RGBIO support_ (unclear if this is due to poor support, or just flawed data)

##### Firmware version
- 75% of users are using 2.5, 2.6 or 2.7 (23%, 30% and 23% respectively)

##### Suggestions
- The most popular suggestion was surrounding improved external interfacing (web / mobile device).
- The second most popular suggestion is to improve sparge control logic.

**Comment by ericyanush** [comment]
Based on the current survey results, I would propose:
- Direct Fired RIMS support is removed (Vessel roles would likely encompass enough of this anyway)
- Old COM Schemas are removed.
- We unify units, as sensors provide measurements in metric, and metric math is simpler, we use solely metric units internally, and have UI code convert to imperial when displaying. 
- We remove the manual volume display option
- The options @dseidman suggested to remove (the set including boil off gallons and buzzer modulation)
- Extract and modularize sparge logic; provides ability to support both fly and batch sparge

**Comment by mrshake** [comment]
I'm good with all of this... but lets make sure we document any features
that are being removed so we can include it in a doc for folks who want to
make the decision to upgrade

**Comment by dseidman** [comment]
Don't necessarily write off options that aren't being used - it may be that
they are unusable at present. For example I very much want a
flowrate-controlled pump, but the relevant options don't even compile for
me. Batch sparge isn't implemented at all.

**Comment by ericyanush** [comment]
I disagree. If an option is currently unusable due to erred implementation or lack thereof, the option is just creating unnecessary complexity and confusion. I would rather see even occasionally used options obliterated from the code base so that we can make the core functionality of the system the absolute best  it can be, then re-implement the options that are/would be used properly, as opposed to walking on egg-shells around fragile implementations of options that are never used.

Based on the data gleaned from the survey, my suggestion for the direction of 3.0 (which by the sem-ver definition introduces breaking changes), is that we take this opportunity to really break things. Based on the results of the survey there are quite a number of features that for whatever reason, aren't widely used, removing them (for now), would allow us to basically start with a blank canvas and re-implement the core in a much more flexible and robust way. From there, we can re-implement individual features without worrying about breaking the world with every line of code changed or added. This approach would ensure that all of the dead code would be removed along with all the unused features, and result in code that we could actually properly test and maintain. It would also give us the change to ensure that everything, from both a developer's and user's perspective is appropriately documented, and to develop some clear style guidelines. 

Thoughts?

**Comment by dseidman** [comment]
I don't think we disagree - I support your approach. What I was getting at
is that we shouldn't _permanently_ cut a feature because it isn't currently
used. I totally agree that we should use the survey to prioritize features,
and we should not feel like we have to reimplement every option that
currently exists, _especially_ ones that don't even work today. So, net, +1
to what you said. Minimum Viable Product, etc. We should make the standard
3 vessel direct fire, RIMS, and HERMS configs and a BIAB config work great.
We might consider preserving low-use options where they are purely a
constant, since implementation and test effort and risk of bugs are so
minimal.

All that said, metered flow control is the last thing between me and a 100%
hands off system, and hop addition delay is critical to make my hop
carousel work, so I'll probably keep those around in my own branch.

One thought: most options did have at least a few people using them. Were
the same few people tweaking many options, or did many people tweak one or
two things? Because if we eliminate a lot of options and everyone loses
something they value, that could be bad.

(This paragraph wound up being a tangent, but it seemed useful enough to
keep.) If you look at the way I structured the sparge logic in my branch,
it makes metered flow control super easy, and any other kind of sparge
logic you might want. Basically there is a FlowController object that
meters the flow between two vessels. Right now it supports only basic PID
flow matching (either to a fixed target or to the other flow rate in the
system), hysteresis-based matching and fill-to-target, but adding batch
sparging would be pretty trivial - except that there's no UI code for it.
Some day I'll even get around to testing my new code so you can integrate
it - but even if I don't I would suggest using it as a start, I tried to
comment it really well so you could tell what I was doing. Let me know if
you want a walkthrough. Basic idea is Vessel objects manage their own
temperature outputs, valves, and volume and the rest of the program just
has to tell them what temperature and volume they should be at.
FlowController objects tell Vessels what volume they should be at, both for
basic fill operations but especially for sparge operations where that
target volume varies over time (the real reason to have FlowControllers in
the first place). There are still some hacks in there which could be
eliminated if we're okay with breaking changes, e.g. having separate valve
profiles for (e.g.) mash in vs. sparge in requires that the vessels know
which stage the program is at, which is unclean. I fixed some hacks in
creating this, e.g. all of the different ways to set on/off/auto all use
the same functions now.

**Comment by ericyanush** [comment]
Since we all are on the same page, we should hammer out a rough plan to move forward with. I'll draft up  a rough outline, and we can make changes as needed. 

To answer your question about users changing options: it wasn't something that I had explicitly checked. But a quick couple of filters on the responses show that roughly half of the users make no changes outside of what I would consider core options (enabling hardware support for expansion boards or enabling brew-step automations). A further 20% of users change only one or two options, which are most commonly the strike temperature offset and the Mash aux temperature sensor averaging. The remaining ~30% of users represent the "power user" category; these are the users that are changing options the majority of the options, and in 99% of cases they are the only ones using them.
TL;DR:
We could remove ~80% of the features and effect only ~30% of the users. Now, this doesn't take into account the fact that some of the options people are changing would likely get rolled into a sweeping redesign (things like the vessel options), so that 30% might even be a tad high. 

That being said, once we had a stable core firmware that would appease the majority of the users, there is no reason we couldn't add back in the popular "power user" features.

**Comment by mrshake** [comment]
I like where you guys are going with this!!!

**Comment by jaimevetter** [comment]
ericyanush, what is the status of the draft outline?

I have finished the changes needed for my system and was going to upstream some of the bugs that I found and cleanup I did. I'm wondering if I should hold off since the suggested changes in this issue may make them irrelevant?

**Comment by jvetter1** [comment]
ericyanush, what is the status of the draft outline?

**Comment by ericyanush** [comment]
@jvetter1 I haven't had a significant amount of time to spend on it yet. My plan is to use [BrewTroller/BrewTroller-Docs](https://github.com/brewtroller/brewtroller-docs) to document the design and anything else that is needed. This way proposals to change the design can be made using pull requests. I should have the first draft of the documentation pushed soon.
